pxr/usd/usd/stageCache.cpp

```diff
@@ -65,69 +58,159 @@ std::atomic_long idCounter(9223000);
-// those elements where pred(element) is true, erase the element from the index
-// and invoke out->push_back(element) if out is not null.  Return the number of
-// elements erased.
-template <class Index, class Range, class Pred, class BackIns>
-size_t EraseIf(Index &index, Range range, Pred pred, BackIns *out) {
-    size_t numErased = 0;
-    while (range.first != range.second) {
-        if (pred(*range.first)) {
-            if (out)
-                out->push_back(*range.first);
-            index.erase(range.first++);
-            ++numErased;
-        } else {
-            ++range.first;
+    size_t EraseAll(const SdfLayerHandle& rootLayer,
+                    const SdfLayerHandle& sessionLayer,
+                    vector<Entry>* erased) {
+        return _EraseAllIf(rootLayer,
+                          [&sessionLayer](const auto& stage){
+                              return stage->GetSessionLayer() == sessionLayer;
+                          },
+                          erased);
+    }
+
+    size_t EraseAll(const SdfLayerHandle& rootLayer,
+                    const SdfLayerHandle& sessionLayer,
+                    const ArResolverContext& resolverContext,
+                    vector<Entry>* erased) {
+        return _EraseAllIf(rootLayer,
+                          [&sessionLayer, &resolverContext](const auto& stage){
+                              return stage->GetSessionLayer() == sessionLayer &&
+                                     stage->GetPathResolverContext() == resolverContext;
+                          },
+                          erased);
+    }
+
+    size_t size() const { return _byId.size(); }
+
+private:
+    // Helper function to erase the ByRootLayer() entry associated with the
+    // stage (and no other entries).
+    //
+    // This must be used in tandem with erase methods on ById() and ByStage()
+    // to uphold the synchronization invariant.
+    void _EraseRootLayerEntry(const UsdStageRefPtr& stage) {
+        const auto range = _byRootLayer.equal_range(
+            stage->GetRootLayer()
+        );
+        const auto it = std::find_if(
+            range.first, range.second,
+            [&stage](const auto& element){
+                return element.second == stage;
+            });
+        if (TF_VERIFY(it != range.second)) {
+            _byRootLayer.erase(it);
         }
     }
-    return numErased;
-}
+
+    template <typename ConditionFn>
+    size_t _EraseAllIf(const SdfLayerHandle& rootLayer,
+                       ConditionFn&& conditionFn,
+                       vector<Entry>* erased) {
+        const auto range = _byRootLayer.equal_range(rootLayer);
+        size_t erasedCount = 0;
+        // In C++20, consider using a filtered range with the condition
+        // function
+        for (auto it = range.first; it != range.second;) {
+            if (conditionFn(it->second)) {
+                const auto byStageIt = _byStage.find(it->second);
```

Comment:
Sorry for nitpicking, but the only issue here is that it this verify does happen to fail, this will loop infinitely since the iterator doesn't get updated. Maybe starting at 200, the code should be
`
                    it = _byRootLayer.erase(it);
                    ++erasedCount;
                    continue;
                }
            } 
            ++it;
`

Reply:
I restructured the loop to avoid the need for a `continue` in the spirit of avoiding in non-trivial loops (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-continue).